### PR TITLE
Iterative approach completed

### PR DESCRIPTION
**Pull request by RBGeomaticsRob**

### CODE REVIEW

File: lib/my_array.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    if arg1.is_a? Symbol
+      return inject_symbol(arg1)
+    elsif arg2.is_a? Symbol
+      return inject_symbol_with_initial(arg1, arg2)
+    elsif arg1.nil?
+      return inject_block(&block)
+    else
+      return inject_block_with_initial(arg1, &block)
+    end
+  end
```
Comment (sjmog) [inline lib/my_array.rb:16]: Beautifully clear OOP

File: lib/my_array.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+class MyArray < Array
```
Comment (sjmog) [inline lib/my_array.rb:1]: Ooooh! You're obeying the Liskov Substitution Principle, get you!

File: lib/my_array.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  def initialize(array)
+    @array = array
+  end
```
Comment (sjmog) [inline lib/my_array.rb:4]: Is this necessary? Can you not just use the implicit `self` to do things like `drop(1)` without referencing state explicitly?

File: lib/my_array.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    elsif arg1.nil?
+      return inject_block(&block)
+    else
+      return inject_block_with_initial(arg1, &block)
+    end
+  end
+
+  private
+
+  attr_reader :array
```
Comment (sjmog) [inline lib/my_array.rb:20]: Boo-urns: these weird me out this far down a class. `attr_reader` should really only declare public interfaces

File: lib/my_array.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  def array_shifted!
+    array.drop(1)
+  end
+
+  def inject_symbol(symbol)
+    memo = array.first
+    array_shifted!.each { |value| memo = memo.send(symbol, value) }
+    memo
+  end
```
Comment (sjmog) [inline lib/my_array.rb:30]: Very nice

File: lib/my_array.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    memo = symbol
+    array.each { |value| memo = memo.send(initial, value) }
+    memo
+  end
+
+  def inject_block(&block)
+    memo = array.first
+    array_shifted!.each { |value| memo = block.call(memo, value) }
+    memo
+  end
```
Comment (sjmog) [inline lib/my_array.rb:42]: I wonder, given the very minor differences between the logic in these methods, if there might have been a DRYer way to shmoosh them all together? Obviously you chose an OOP approach here, which is laudable.

File: lib/my_array.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+  def inject_block_with_initial(initial, &block)
+    memo = initial
+    array.each { |value| memo = block.call(memo, value) }
```
Comment (sjmog) [inline lib/my_array.rb:46]: Lovely metaprogramming and use of `to_proc` (`&`)

File: spec/my_array_spec.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    expect(array.inject(:+)).to eq 15
+    expect(subject.inject(:+)).to eq 15
+  end
+
+  it 'can multiply elements in an array using a symbol and a start point' do
+    expect(array.inject(5, :*)).to eq 600
+    expect(subject.inject(5, :*)).to eq 600
+  end
+
+  let(:string_array) { %w(A b C) }
```
Comment (sjmog) [inline spec/my_array_spec.rb:26]: Beautiful setting up

File: spec/my_array_spec.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
+require 'my_array'
+
+describe MyArray do
+  let(:array) { (1..5) }
+  let(:subject) { described_class.new(array) }
```
Comment (sjmog) [inline spec/my_array_spec.rb:5]: Did you...did you just use the `described_class` keyword? Awesome!

Comment (sjmog) [inline spec/my_array_spec.rb:5]: Bear in mind you don't have to `let(:subject)` (and should avoid it) - doing

``` ruby
subject { described_class.new(array) }
```

would do.

File: spec/my_array_spec.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    expect(string_array.inject(:+)).to eq "AbC"
+    expect(string_subject.inject(:+)).to eq "AbC"
+  end
+
+  it 'can accept string values in the array with a block' do
+    expect(string_array.inject { |sum, item| sum += item }).to eq "AbC"
+    expect(string_subject.inject { |sum, item| sum += item }).to eq "AbC"
+  end
+
+end
```
Comment (sjmog) [inline spec/my_array_spec.rb:39]: Lovely spec file

**Comment by sjmog** [comment]
A beautiful solution here, Rob - 'nuff said. Great use of OOP principles :+1: and spec helpers. Love it. 

:checkered_flag:

File: lib/my_array.rb (hunk 4 above)
Comment (RBGeomaticsRob) [inline lib/my_array.rb:20]: Thanks Sam, I had been using them to DRY out the @'s without providing a reader to the API, point taken.

File: lib/my_array.rb (hunk 3 above)
Comment (RBGeomaticsRob) [inline lib/my_array.rb:4]: Yes, what I hadn't taken into account here is that by inheriting from array, I already had the ability to take in an argument on the .new method.

**Pull request closed without merging**